external/fv3fit/fv3fit/keras/_models/convolutional.py

```diff
@@ -81,20 +84,28 @@ def get_Xy_dataset(
     requested input variables and the second containing
     the requested output variables.
     """
-    # tile, x, y, z
-    data = data.map(apply_to_mapping(ensure_nd(4)))
+    # sample, tile, x, y, z
+    data = data.map(apply_to_mapping(ensure_nd(5)))
+
     if clip_config is not None:
-        y_source = data.map(clip_sample(clip_config))
+        clip_function = clip_sample(clip_config)
     else:
-        y_source = data
-    y = y_source.map(select_keys(output_variables))
-    X = data.map(apply_to_mapping(append_halos_tensor(n_halo))).map(
-        select_keys(input_variables)
-    )
-    # now that we have halos, we need to collapse tile back into the sample dimension
-    X = X.unbatch()
-    y = y.unbatch()
-    return tf.data.Dataset.zip((X, y))
+
+        def clip_function(data):
+            return data
+
+    def map_fn(data):
+        # clipping of inputs happens within the keras model, we don't clip at the
+        # data layer so that the model still takes full-sized inputs when used
+        # in production
+        x = apply_to_tuple(append_halos_tensor(n_halo))(
+            select_keys(input_variables, data)
+        )
+        y = select_keys(output_variables, clip_function(data))
+        return x, y
```

Comment:
This pattern suggested by @nbren12 was already used in the dense code in the original tfdataset PR, adding it to convolutional training here.